Sephora-M/assignment3/LetterGrid.java

```diff
@@ -0,0 +1,117 @@
+package assignment3;
+
+import java.awt.Point;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+
+/**Grid of letters. Contains a method that, given a dictionnary represented as a Trie, allows to find 
+ * all the words that can be formed in the grid. 
+ * 
+ * @author Sephora-M
+ */
+public class LetterGrid {
+	char[][] grid;
+	int X;
+	int Y;
+	
+	public LetterGrid(char[][] grid) {
+		if (grid == null || grid[0].length == 0)
+			throw new IllegalArgumentException("Cannot instatiate an empty or a null grid.");
+		this.grid = grid;
+		Y = grid.length;
+		X = grid[0].length;
+	}
+	
+	/**
+	 * Given a Trie object, returns a set of all the words in the Trie that can be formed in the grid 
+	 * @param dict a Trie, the dictionary
+	 * @return set of all the words from the dictionary that can be from in the grid
+	 * 
+	 * Note: the idea is to start from each letter in the grid and perform a DFS-like search. To optimize
+	 * the search, check at each time if the current word that we are constructing is a prefix of a word in the 
+	 * dictionary. It is not exactly a DFS, the subtility here is that we don't just want to visit all the 
+	 * nodes, but we want to visit them from different path. This is taken care here by keeping track of a
+	 * depthMap, which tells us when we visited each node. 
+	 */
+	public Set<String> dictionnaryWords(Trie dict){
+		HashSet<String> foundWords = new HashSet<String>();
+
+		for(int x = 0; x < X; x++) {
+			for(int y = 0; y < Y; y++) {
+				int[][] depthMap = new int[Y][X];
+				int currentDepth = 1;
+				depthMap[y][x] = currentDepth;
```

Comment:
If you do a DFS recursively from each node, mark a node visited when you enter it in searchWords (if it hasn't been visited yet) and then unmark it as visited at the end of searchWords, then you can just use one boolean[][] array here, and reuse it instead of doing deep copies. This would save a lot of space for even small grids.

Reply:
Ahh I was scratching my head to find this solution.. I didn't think of simply setting it back to unvisited at the end of searchWord.. I love this solution, I've adapted the code accordingly, thanks a lot!